File: PrimaMateria/PrimaMateria/Classes/DynoTableHeaderButton.m
//
//  DynoTableHeaderButton.m
//  PrimaMateria
//
//  Created by Jerry Porter on 1/5/2011.
//  Copyright (c) 2016 xTrensa. All rights reserved.
//

@implementation DynoTableHeaderButton

- (void)tableHeaderCellSelected: (NSNotification *)aNotification {
    self.imageLayer.contents = nil;
}

- (void)awakeFromNib {
    [super awakeFromNib];
    
	[[NSNotificationCenter defaultCenter] addObserver: self selector: @selector(tableHeaderCellSelected:) name: NOTIFICATION_TABLE_HEADER_SELECTED object: nil];
    self.toggle = 0;
    self.ascendingImageView = [UIImage imageNamed: @"AscendingArrow.png"];
    self.descentingImageView = [UIImage imageNamed: @"DescendingArrow.png"];
    self.imageLayer = [CALayer layer];
    [self.layer addSublayer: self.imageLayer];
    self.backgroundColor = UIColor.lightGrayColor;
}

- (void)toggleButtonState: (BOOL) aFlag {
    [[NSNotificationCenter defaultCenter] postNotificationName:NOTIFICATION_TABLE_HEADER_SELECTED object: nil];
    [CATransaction begin];
    [CATransaction setAnimationDuration:0.0];
    self.imageLayer.frame = (aFlag) ? CGRectMake((self.frame.size.width/2) - 6, 2, 12, 12) : CGRectMake((self.frame.size.width/2) - 6, 30, 12, 12);
    self.imageLayer.contents = (aFlag) ? (id)self.ascendingImageView.CGImage : (id)self.descentingImageView.CGImage;
    [CATransaction commit];
    self.toggle = aFlag;
}

- (BOOL)toggleState {
    [self toggleButtonState: !self.toggle];
    return self.toggle;
}

- (void)createLayer {
    self.layer.masksToBounds = YES;
    self.layer.borderWidth = 1.0f;
}

- (void)dealloc {
    [[NSNotificationCenter defaultCenter] removeObserver: self];
}

@end

File: PrimaMateria/PrimaMateria/Classes/DynoTableHeaderButton.swift
//
//  DynoTableHeaderButton.swift
//  PrimaMateria
//
//  Created by Jerry Porter on 4/14/16.
//  Copyright © 2016 xTrensa. All rights reserved.
//

@objc class DynoTableHeaderButton : GeneralButton {
    var imageLayer : CALayer?
    var descentingImageView : UIImage?
    var ascendingImageView : UIImage?
    var toggle : Bool = false
        
    // MARK: - Initialization Methods
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder:aDecoder)
    }
    
    override func awakeFromNib() {
        super.awakeFromNib()
        
        NSNotificationCenter.defaultCenter().addObserver(self, selector: #selector(DynoTableHeaderButton.tableHeaderCellSelected(_:)), name: NOTIFICATION_TABLE_HEADER_SELECTED, object: nil)
        self.toggle = false
        self.ascendingImageView = UIImage.init(named: "AscendingArrow.png")
        self.descentingImageView = UIImage.init(named: "DescendingArrow.png")
        self.imageLayer = CALayer.init()
        self.layer.addSublayer(self.imageLayer!)
        self.backgroundColor = UIColor.lightGrayColor()
    }
    
    // MARK: - Notification Methods
    
    func tableHeaderCellSelected(aNotification: NSNotification) {
        self.imageLayer!.contents = nil
    }

    // MARK: - Internal Methods

    func toggleButtonState(aFlag: Bool) {
        NSNotificationCenter.defaultCenter().postNotificationName(NOTIFICATION_TABLE_HEADER_SELECTED, object: nil)
        CATransaction.begin()
        CATransaction.setAnimationDuration(0.0)
        self.imageLayer!.frame = (aFlag) ? CGRectMake((self.frame.size.width/2) - 6, 2, 12, 12) : CGRectMake((self.frame.size.width/2) - 6, 30, 12, 12)
        self.imageLayer!.contents = (aFlag) ? self.ascendingImageView!.CGImage : self.descentingImageView!.CGImage
        CATransaction.commit()
        self.toggle = aFlag
    }
    
    func toggleState() -> Bool {
        self.toggleButtonState(!self.toggle)
        return self.toggle
    }
    
    override func createLayer() {
        self.layer.masksToBounds = true
        self.layer.borderWidth = 1.0
    }
    
    deinit {
        NSNotificationCenter.defaultCenter().removeObserver(self)
    }
}
